maven/lib/dependabot/maven/file_updater/declaration_finder.rb

```diff
@@ -11,7 +11,7 @@ class FileUpdater
       class DeclarationFinder
         DECLARATION_REGEX =
           %r{<parent>.*?</parent>|<dependency>.*?</dependency>|
-             <plugin>.*?</plugin>|<extension>.*?</extension>}mx.freeze
+             <plugin>.*?(?:<plugin>.*?</plugin>.*)?</plugin>|<extension>.*?</extension>}mx.freeze
```

Comment:
I saw that Ruby has the ability to express [recursive regex](https://ruby-doc.org/core-2.7.0/Regexp.html#class-Regexp-label-Subexpression+Calls). I'm in the process of wrapping my mind around that, but I'm wondering if we could use that functionality. Are you familiar?

If you think recursive regex is a promising solution, I think it's worth spending some time to get it right (and to make it as readable as possible). If not, or if it seems too esoteric, I think what you have added here will solve the problem that is in front of us 👍

Reply:
I've not had to use it, could you show an example of what the regex could look like and what edgecases it would catch over the current implementation? It doesn't have to fully work, but just to express your thinking?

Reply:
> could you show an example of what the regex could look like

I've been trying to grok the syntax for recursive regex. I'm not great at recursive thinking, and handling multiple tag types makes that more of a challenge. After trying this out in Rubular for a bit, I'm ready to say the investment isn't worth it for now.

>  what edgecases it would catch over the current implementation

My thinking was that we'd be able to possibly catch the edge case we saw with nested `<plugin>` declarations for each of the tags we're looking at. I was less concerned with edge cases and more concerned with not having to handle other issues as they pop up (e.g. nested `<extension>` tags).

---

Conclusion here is that I don't think the investment is worth it for now.